implement preserve.replicas when restaging files

cookie33 opened the pull request:
Implement fix for: https://github.com/irods/irods_capability_storage_tiering/issues/233

alanking reviewed (commented):
Yep, looks good. We (iRODS team) will discuss how we want to approach adding an automated test for this (as @trel described in https://github.com/irods/irods_capability_storage_tiering/issues/233#issuecomment-1850113361). In the meantime, feel free to take a swing at it yourself. ;)

alanking reviewed (commented):
I went ahead and wrote a test for this and may have discovered an issue with the fix. Please confirm that suggested fix works for your use case.

cookie33 commented on storage_tiering.cpp:
```diff
@@ -817,7 +817,7 @@ namespace irods {
                 low_tier_resource_name,
                 get_verification_for_resc(comm_, low_tier_resource_name),
-                false,
+                get_preserve_replicas_for_resc(comm_, low_tier_resource_name),
```
> Actually... should this be `_source_resource` instead of `low_tier_resource_name`?
> 
> We are intending to preserve the replica on the resource _from which_ we are restaging (i.e. source resource), so it seems that we should check the configuration on that resource rather than the minimum restage tier (i.e. destination resource).

Hi,
You might be right.
But the fix breaks the configuration. :(
If the parameter is not set it will now give an error and do nothing at all. No restaging of objects to an higher tier.
But it will get the object directly from the lower tier.

Will implement a better fix.

alanking replied:
@cookie33 - I'm guessing that you got this working?

cookie33 replied:
Yup. This works as advertised in the code.
And that is why I have test: `test_preserve_replicas_works_with_restage_when_replica_only_exists_in_last_tier__issue_233`

alanking replied:
Cool, thanks!

cookie33 commented:
Changed one query:
```
-            boost::format("SELECT META_DATA_ATTR_VALUE WHERE DATA_NAME = '%s' AND COLL_NAME = '%s' AND META_DATA_ATTR_NAME = '%s' AND META_DATA_ATTR_UNITS = '%s'")
+            boost::format("SELECT META_DATA_ATTR_VALUE WHERE DATA_NAME = '%s' AND COLL_NAME = '%s' AND META_DATA_ATTR_NAME = '%s' AND META_DATA_ATTR_UNITS >= '%s'")
```
This makes sure that the function `get_group_name_by_replica_number` will work even if the replica has a lower number than the unit in `irods::storage_tiering::group`

alanking commented:
I left a comment in #232 regarding the query update: https://github.com/irods/irods_capability_storage_tiering/issues/232#issuecomment-1852808480

cookie33 commented:
I made it a draft pull request.
I will make some tests in bash and when that is OK I will convert it back to a real pull request again.
The test will be similar like the automated tests from the irods consortium itself.

korydraughn commented:
`packaging/test_plugin_unified_storage_tiering.py` contains all of our tests. All tests are expected to be implemented using Python.

We can also translate your bash tests into python if that helps.

cookie33 commented:
I tried running the python tests on my system. But the first thing it does is stop the iRODS server :(

I will try to get it running in a similar fashion as the iRODS tests. This will make it easier to incorporate.
After that I will update the official `test_plugin_unified_storage_tiering.py` with my updates after testing as much as possible.

korydraughn commented:
That's correct. The `run_tests.py` script restarts the server before launching the tests to make sure the environment is ready.

Tests which depend on the delay server (like this project) normally restart the server too.

We're happy to help you cross the finish line :-)

korydraughn commented:
Forgot to mention that you can use the iRODS Development Environment to run the tests easier. See https://github.com/irods/irods_development_environment.

With that, you can use the plugin builder and plugin test runners to build and test.

cookie33 commented:
Updated `test_plugin_unified_storage_tiering_test .py` in the class `TestStorageTieringPluginPreserveReplica`

Renamed one test to `test_put` because it did not do a get.
Added 4 other tests where preserve_replicas is set to true.

tested:
```
$ time python3 run_tests.py --run_specific_test test_plugin_unified_storage_tiering > /tmp/test.txt
irods.test.test_plugin_unified_storage_tiering_test.TestStorageTieringPluginPreserveReplica.test_put ... ok
irods.test.test_plugin_unified_storage_tiering_test.TestStorageTieringPluginPreserveReplica.test_put_and_get_01 ... ok
irods.test.test_plugin_unified_storage_tiering_test.TestStorageTieringPluginPreserveReplica.test_put_and_get_02 ... ok
irods.test.test_plugin_unified_storage_tiering_test.TestStorageTieringPluginPreserveReplica.test_put_and_get_03 ... ok
irods.test.test_plugin_unified_storage_tiering_test.TestStorageTieringPluginPreserveReplica.test_put_and_get_04 ... ok

----------------------------------------------------------------------
Ran 5 tests in 308.318s

OK

real    5m8.492s
user    0m4.896s
sys     0m1.739s
```

alanking reviewed (commented):
Thanks for writing these tests! I think that the first test you wrote covers the ground and we probably don't need the other tests. I could be wrong, so please let me know if I misunderstood anything.

alanking commented on storage_tiering.cpp:
```diff
@@ -763,7 +763,7 @@ namespace irods {
 
         std::string qstr{boost::str(
-            boost::format("SELECT META_DATA_ATTR_VALUE WHERE DATA_NAME = '%s' AND COLL_NAME = '%s' AND META_DATA_ATTR_NAME = '%s' AND META_DATA_ATTR_UNITS = '%s'")
+            boost::format("SELECT META_DATA_ATTR_VALUE WHERE DATA_NAME = '%s' AND COLL_NAME = '%s' AND META_DATA_ATTR_NAME = '%s' AND META_DATA_ATTR_UNITS >= '%s'")
```
I think we should just remove the `META_DATA_ATTR_UNITS` filter, as you suggested here: https://github.com/irods/irods_capability_storage_tiering/issues/232#issuecomment-1851900106
```suggestion
            boost::format("SELECT META_DATA_ATTR_VALUE WHERE DATA_NAME = '%s' AND COLL_NAME = '%s' AND META_DATA_ATTR_NAME = '%s'")
```

cookie33 replied:
Hi,
There are two ways to implement this:
* only change the sql query (very easy, but not complete)
* change the function `storage_tiering::get_group_name_by_replica_number` to not have the parameter `_replica_number` and change the sql query. This change means you have to fix all calls to the function.

trel replied:
I believe we never want to depend on the specifics of replica numbers being in a particular layout / usage pattern (assuming replica `0`, or that one is greater/less than another, etc.).

So I think we must go with option B (and with it, lose the possibility of a data object being in multiple tier groups at the same time).

Discussion is still open of course.

alanking replied:
Right, replica numbers only go up so this seems like it would work. However, it makes me a little bit nervous for the reason @trel laid out, and doesn't really do anything more than just removing the replica number filter for our existing supported use cases.

Fortunately, I see only 2 uses of this function, so replacing it should be easy. I think your suggestion of renaming the function or making a new function called `get_group_name_for_object` or something along those lines is a good one.

cookie33 replied:
I am OK with option 2
If everybody agrees I will implement it somewhere next week. Just let me know.

alanking replied:
Agreed. Thanks!

cookie33 replied:
Added a new commit which renames `get_group_name_by_resource_number` to `get_group_name_for_object`.
Excluded ` AND META_DATA_ATTR_UNITS = '%s'` from sql query.
Removed parameter `_replica_number` from the function.

Tested locally OK.

alanking replied:
Excellent. Resolving

cookie33 commented on packaging/test_plugin_unified_storage_tiering.py:
```diff
@@ -776,6 +776,176 @@ def test_put_and_get(self):
                 finally:
                     admin_session.assert_icommand('irm -f ' + filename)
 
+    def test_put_and_get_01(self):
```
Ok. Changed name

cookie33 commented on packaging/test_plugin_unified_storage_tiering.py:
```diff
@@ -776,6 +776,176 @@ def test_put_and_get(self):
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '2 ufs0')
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '1 ufs1')
+
+                    # test restage file from tier 1 again
+                    admin_session.assert_icommand('iget ' + filename + ' - ', 'STDOUT_SINGLELINE', 'TESTFILE')
+
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '2 ufs0')
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '1 ufs1')
```
OK. Removed test

cookie33 commented on packaging/test_plugin_unified_storage_tiering.py:
```diff
@@ -776,6 +776,176 @@ def test_put_and_get(self):
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '2 ufs2')
+
+                    # test restage to tier 1
+                    admin_session.assert_icommand('iget ' + filename + ' - ', 'STDOUT_SINGLELINE', 'TESTFILE')
```
OK. Changed test

cookie33 commented on packaging/test_plugin_unified_storage_tiering.py:
```diff
@@ -776,6 +776,176 @@ def test_put_and_get(self):
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '1 ufs1')
+
+                    # remove the object in tier 0
+                    admin_session.assert_icommand('itrim -N1 -n0 ' + filename, 'STDOUT_SINGLELINE', 'Number of files trimmed = 1' )
```
This shows the problem of issue 233 I saw.
I like it because it demostrates the way we (I) want to implement the storage tiering with two tiers.
Both have preserve_replicas set to true. And we remove an object using itrim from tier 0 if space is needed using a seperate reoccuring rule.
I changed the name to: `test_preserve_replicas_works_with_restage_when_replica_only_exists_in_last_tier__issue_233`

alanking replied:
Ah, okay. The updated name helped me see what's going on. Thanks!

cookie33 commented on packaging/test_plugin_unified_storage_tiering.py:
```diff
@@ -776,6 +776,176 @@ def test_put_and_get(self):
+                    admin_session.assert_icommand('itrim -N1 -n0 ' + filename, 'STDOUT_SINGLELINE', 'Number of files trimmed = 1' )
+
+                    # test restage file from tier 1. Causes restage to tier 0
+                    admin_session.assert_icommand('iget -R ufs1 ' + filename + ' - ', 'STDOUT_SINGLELINE', 'TESTFILE')
```
Ok. Removed

alanking reviewed (commented):
Looking very good. Saw one trailing space that I'm guessing was not on purpose.

I think we are almost ready to squash this. @cookie33 - Are you comfortable squashing this down to 1 or 2 commits? I would suggest 1 commit for #232 and 1 commit for #233 with commit messages like this:
```
[#232] My super helpful commit title

And maybe a description with the what and why, if you're into it.
```

If you're busy, we can swizzle the commits (your authorship will be retained). Up to you!

cookie33 commented on storage_tiering.cpp:
```diff
@@ -753,33 +753,31 @@ namespace irods {
             THROW(
                 CAT_NO_ROWS_FOUND,
-                "failed to fetch group name by resource and replica number");
+                "failed to fetch group name by object and resource ");
```
Hi Alan,
Can you swizzle the commits. At the moment I am busy trying to go from standard iRODS rules to iRODS python based rules and my `normal` system admin tasks.
I did fix the extra space at the end of the error message.

alanking reviewed (approved):
Commits swizzled locally. Build succeeded. Running tests. Once these have passed, I'll put this in.

Great work, @cookie33!

alanking commented:
Confirmed that tests are passing. Added #s. Merging.
